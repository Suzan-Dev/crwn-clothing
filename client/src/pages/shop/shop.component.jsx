import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { fetchCollectionStart } from '../../redux/shop/shop.action';
import { Outlet } from 'react-location';

export const ShopPage = ({ fetchCollectionStart }) => {
  useEffect(() => {
    fetchCollectionStart();
  }, [fetchCollectionStart]);

  return (
    <div className="shoppage">
      <Outlet />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  fetchCollectionStart: () => dispatch(fetchCollectionStart()),
});

export default connect(null, mapDispatchToProps)(ShopPage);
