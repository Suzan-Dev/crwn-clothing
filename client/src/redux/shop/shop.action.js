import ShopActionTypes from './shop.type';

export const fetchCollectionStart = () => ({
  type: ShopActionTypes.FETCH_COLLECTION_START,
});

export const fetchCollectionSuccess = (collections) => ({
  type: ShopActionTypes.FETCH_COLLECTION_SUCCESS,
  payload: collections,
});

export const fetchCollectionFailure = (errorMsg) => ({
  type: ShopActionTypes.FETCH_COLLECTION_FAILURE,
  payload: errorMsg,
});
