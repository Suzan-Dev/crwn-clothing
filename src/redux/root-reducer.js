import { combineReducers } from 'redux';

import UserReducer from './user/user.reducer';
import CartReducer from './cart/cart.reducer';

// this is Redux Store -> which is an object
export default combineReducers({
  user: UserReducer,
  cart: CartReducer,
});
