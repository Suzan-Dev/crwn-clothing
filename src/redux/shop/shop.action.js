import ShopActionTypes from './shop.type';

import {
  firestore,
  convertCollectionToShopData,
} from '../../firebase/firebase.utils';

export const fetchCollectionStart = () => ({
  type: ShopActionTypes.FETCH_COLLECTION_START,
});

export const fetchCollectionSuccess = (collections) => ({
  type: ShopActionTypes.FETCH_COLLECTION_SUCCESS,
  payload: collections,
});

export const fetchCollectionFailure = (errorMsg) => ({
  type: ShopActionTypes.FETCH_COLLECTION_FAILURE,
  payload: errorMsg,
});

export const fetchCollectionStartAsync = () => {
  return (dispatch) => {
    const collectionRef = firestore.collection('collection');
    dispatch(fetchCollectionStart());

    // Observer Pattern
    collectionRef
      .get()
      .then(async (snapshot) => {
        const updatedCollection = convertCollectionToShopData(snapshot);
        dispatch(fetchCollectionSuccess(updatedCollection));
      })
      .catch((err) => dispatch(fetchCollectionFailure(err.message)));
  };
};
