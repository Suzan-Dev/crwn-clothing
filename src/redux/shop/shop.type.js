const ShopActionTypes = {
  UPDATED_COLLECTION: 'UPDATED_COLLECTION',
};

export default ShopActionTypes;
