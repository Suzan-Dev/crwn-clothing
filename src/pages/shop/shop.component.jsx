import React from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

// Container Pattern
import CollectionOverviewContainer from '../../components/collection-overview/collection-overview.container';
import CollectionPageContainer from '../collection/collection.container';

import { fetchCollectionStart } from '../../redux/shop/shop.action';

class ShopPage extends React.Component {
  componentDidMount() {
    const { fetchCollectionStart } = this.props;
    fetchCollectionStart();
  }

  render() {
    const { match } = this.props;
    return (
      <div className='shoppage'>
        <Route
          exact
          path={`${match.path}`}
          component={CollectionOverviewContainer}
        />
        <Route
          exact
          path={`${match.path}/:collectionId`}
          component={CollectionPageContainer}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  fetchCollectionStart: () => dispatch(fetchCollectionStart()),
});

export default connect(null, mapDispatchToProps)(ShopPage);
