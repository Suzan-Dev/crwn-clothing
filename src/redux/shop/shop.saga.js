import { takeEvery, call, put, all } from 'redux-saga/effects';

import ShopActionTypes from './shop.type';

import {
  firestore,
  convertCollectionToShopData,
} from '../../firebase/firebase.utils';

import { fetchCollectionSuccess, fetchCollectionFailure } from './shop.action';

// saga workers
function* fetchCollectionAsync() {
  try {
    const collectionRef = firestore.collection('collection');
    const snapshot = yield collectionRef.get();
    const updatedCollection = yield call(convertCollectionToShopData, snapshot);
    yield put(fetchCollectionSuccess(updatedCollection));
  } catch (err) {
    yield put(fetchCollectionFailure(err.message));
  }
}

// saga watchers
function* fetchCollectionStart() {
  yield takeEvery(ShopActionTypes.FETCH_COLLECTION_START, fetchCollectionAsync);
}

export function* shopSagas() {
  yield all([call(fetchCollectionStart)]);
}
