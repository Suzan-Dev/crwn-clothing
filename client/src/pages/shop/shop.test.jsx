import React from 'react';
import { mount } from '../../setupTests';
import { combineReducers, createStore } from 'redux';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { ShopPage } from './shop.component';

export const createMockStore = ({ state, reducers }) => {
  const store = createStore(combineReducers(reducers), state);
  return {
    ...store,
    persistor: {
      persist: () => null,
    },
  };
};

describe('Testing: ShopPage component!', () => {
  let wrapper;
  let mockFetchCollectionStart;
  let store;

  beforeEach(() => {
    const mockReducer = (
      state = {
        isFetching: true,
      },
      action
    ) => state;

    const mockState = {
      shop: {
        isFetching: true,
      },
    };

    mockFetchCollectionStart = jest.fn();

    store = createMockStore({
      state: mockState,
      reducers: { shop: mockReducer },
    });

    const mockMatch = {
      path: '',
    };

    const mockProps = {
      match: mockMatch,
      fetchCollectionStart: mockFetchCollectionStart,
    };

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ShopPage {...mockProps} />
        </Provider>
      </BrowserRouter>
    );
  });

  test('should render ShopPage component.', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should call fetchCollectionStart on mount.', () => {
    expect(mockFetchCollectionStart).toHaveBeenCalled();
  });
});
