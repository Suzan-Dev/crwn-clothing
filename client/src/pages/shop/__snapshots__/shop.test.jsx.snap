// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: ShopPage component! should render ShopPage component. 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <Provider store={{...}}>
      <ShopPage match={{...}} fetchCollectionStart={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [ [] ], instances: [ undefined ], invocationCallOrder: [ 1 ], results: [ { type: 'return', value: undefined } ] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
        <div className=\\"shoppage\\">
          <Suspense fallback={{...}}>
            <Route exact={true} path=\\"\\" component={{...}} />
            <Route exact={true} path=\\"/:collectionId\\" component={{...}} />
            <Spinner>
              <styled.div>
                <div className=\\"sc-bdnylx gRtvuV\\">
                  <styled.div>
                    <div className=\\"sc-gtssRu gevvoa\\" />
                  </styled.div>
                </div>
              </styled.div>
            </Spinner>
          </Suspense>
        </div>
      </ShopPage>
    </Provider>
  </Router>
</BrowserRouter>"
`;
